### PR TITLE
Feature/device

### PR DESCRIPTION
**Pull request by maddyscientist**
The PR continues the abstraction effort and cleanup some old experiments:
* Move CUDA specific initialization and destruction from `initQuda` and `endQuda`, moving to new target-specific file lib/targets/cuda/device.cpp
* Abstracted `cudaHostGetDevicePointer` to new function, `get_mapped_device_pointer` in malloc.cpp
* In preparation of merging NVSHMEM, remove old asynchronous dslash policy experiments

Notably this PR removes all CUDA specifics from interface_quda.cpp

### CODE REVIEW

File: lib/dslash_policy.cuh [hunk 1]
```diff
@@ -1862,8 +1603,6 @@ namespace quda
     QUDA_FUSED_ZERO_COPY_PACK_GDR_RECV_DSLASH,
     QUDA_DSLASH_FUSED_PACK,
     QUDA_DSLASH_FUSED_PACK_FUSED_HALO,
-    QUDA_DSLASH_ASYNC,
-    QUDA_FUSED_DSLASH_ASYNC,
     QUDA_DSLASH_NC,
```
Comment (weinbe2) [inline lib/dslash_policy.cuh:1606]: Are there any hardcoded comparisons of dslash policies with integers that may bork because the integer value of `QUDA_DSLASH_NC` has changed?

**Comment by maddyscientist** [comment]
Nope

File: lib/interface_quda.cpp [hunk 2]
```diff
@@ -289,28 +267,28 @@ static void profilerStart(const char *f) {
   static int target_count = 0;
   static unsigned int i = 0;
   if (do_not_profile_quda){
-    cudaProfilerStop();
+    device::profile::stop();
```
Comment (weinbe2) [inline lib/interface_quda.cpp:270]: I'm a little confused about why this logic exists in the first place... if we're not profiling QUDA, why do we need to "stop" profiling QUDA in `profilerStart`? I could see a world where this maybe lets users update `QUDA_DO_NOT_PROFILE` from the environment in the middle of a run, but I still don't see how the logic of "stopping" when you're requesting a start makes sense.

File: lib/interface_quda.cpp [hunk 3]
```diff
@@ -289,28 +267,28 @@ static void profilerStart(const char *f) {
...
     }
   }
 }
 }
 
 static void profilerStop(const char *f) {
-  if(do_not_profile_quda){
-    cudaProfilerStart();
+  if (do_not_profile_quda){
+    device::profile::start();
```
Comment (weinbe2) [inline lib/interface_quda.cpp:286]: Same question as above, just flipped, why is this here in the first place?

**Comment by weinbe2** [comment]
A few questions, but otherwise this PR looks reasonable. Happy to see all explicit CUDA references are gone from `interface_quda.cpp`. What testing has been done? (Though I guess this is a pretty compact "you'll know if you messed up quickly at runtime" PR.)

**Review by cpviolator — APPROVED** [review]
Evan raises some good points. As for the dslash policy enumeration, I know that we have some scripts that enable only a subset of policies to save on tuning time, but none of them go as high as the `NC` policy.

File: lib/dslash_policy.cuh (hunk 1 above)
Comment (maddyscientist) [reply lib/dslash_policy.cuh:1606]: Nope, and in fact `QUDA_DSLASH_NC` isn't used at all any more and is another legacy policy that could be deleted (It's from the original Möbius implementation that Hyung-Jin wrote).  I figured it could have some use at some point to have a no-comms policy though, so kept it for now.

**Comment by maddyscientist** [comment]
Thanks for the reviews folks.  

@weinbe2 testing done is is both single and multi-GPU, and also with `QUDA_NVML` enabled.
@cpviolator the policy enumeration hasn't been changed for any polices that are actually used by anyone, so it won't affect any scripts.

**Comment by weinbe2** [comment]
I'm going to test this with Chroma+QDPJIT, hold off on merging it at least until I give it a pass there.

**Review by weinbe2 — APPROVED** [review]
Passes the sanity tests. Good to merge.
